From: Hannes Reinecke <hare@suse.de>
Date: Fri, 11 May 2018 12:10:13 +0200
Subject: [PATCH] block: sanity check for integrity intervals
References: bsc#1091728
Patch-Mainline: submitted linux-block 2018/05/11

When registering for block integrity the template is supposed to
set an 'intervals_exp' value, which needs to be larger than 9.
So we should add a sanity check here to avoid miscalculation
of the integrity payload.

Signed-off-by: Hannes Reinecke <hare@suse.com>
---
 block/bio-integrity.c | 2 ++
 block/blk-integrity.c | 6 ++++--
 2 files changed, 6 insertions(+), 2 deletions(-)

diff --git a/block/bio-integrity.c b/block/bio-integrity.c
index 2cffb4b017cb..0cc6a098f9be 100644
--- a/block/bio-integrity.c
+++ b/block/bio-integrity.c
@@ -285,6 +285,8 @@ int bio_integrity_prep(struct bio *bio)
 	BUG_ON(bi == NULL);
 	BUG_ON(bio_integrity(bio));
 
+	if (WARN_ON(bi->interval_exp < 9))
+		bi->interval_exp = 9;
 	intervals = bio_integrity_intervals(bi, bio_sectors(bio));
 
 	/* Allocate kernel buffer for protection data */
diff --git a/block/blk-integrity.c b/block/blk-integrity.c
index e4ebd79de679..58b906379324 100644
--- a/block/blk-integrity.c
+++ b/block/blk-integrity.c
@@ -412,8 +412,10 @@ void blk_integrity_register(struct gendisk *disk, struct blk_integrity *template
 
 	bi->flags = BLK_INTEGRITY_VERIFY | BLK_INTEGRITY_GENERATE |
 		template->flags;
-	bi->interval_exp = template->interval_exp ? :
-		ilog2(queue_logical_block_size(disk->queue));
+	if (template->interval_exp > 9)
+		bi->interval_exp = template->interval_exp;
+	else
+		bi->interval_exp = ilog2(queue_logical_block_size(disk->queue));
 	bi->profile = template->profile ? template->profile : &nop_profile;
 	bi->tuple_size = template->tuple_size;
 	bi->tag_size = template->tag_size;
-- 
2.12.3

